### PR TITLE
Bug 799615 - send tab from Twitter shares

### PR DESCRIPTION
**Pull request by ncalexan**
This is a heuristic, and it might not be the right heuristic, but it's simple.  r=, rnewman?

### CODE REVIEW

**Comment by rnewman** [comment]
Furthermore: maybe even `findFirstWebURL`. We absolutely do not want to allow a `javascript:` or `file:` URI to be sent, for example. So let's take a test for that.

Otherwise, looks good!

**Comment by ncalexan** [comment]
Reworked this to have a stand-alone WebURLFinder.  Re-review, please.

File: src/main/java/org/mozilla/gecko/sync/setup/activities/SendTabData.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      StringBuffer sb = new StringBuffer();
+      for (String data : new String[] { text, subject, title }) {
+        if (data != null) {
+          sb.append(data);
+          sb.append("\n\n");
+        }
+      }
+
+      // Find first URL.
+      theUri = new WebURLFinder(sb.toString()).bestWebURL();
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/setup/activities/SendTabData.java]: Why not just have WebURLFinder take a collection? Cheaper than concatenating a new string.

File: src/main/java/org/mozilla/gecko/sync/setup/activities/SendTabData.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+package org.mozilla.gecko.sync.setup.activities;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import android.content.Intent;
+import android.os.Bundle;
+
+public class SendTabData {
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/setup/activities/SendTabData.java]: Small comment.

File: src/main/java/org/mozilla/gecko/sync/setup/activities/WebURLFinder.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+public class WebURLFinder {
+  public final List<String> candidates;
+
+  public WebURLFinder(String string) {
+    if (string == null) {
+      throw new IllegalArgumentException("string must not be null");
+    }
+
+    List<String> strings = new ArrayList<String>();
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/setup/activities/WebURLFinder.java]: But there's a `candidateWebURLs` method that takes a single string! Just use that!

**Comment by rnewman** [comment]
r+

**Pull request merged**